PORTING.md

```diff
@@ -0,0 +1,128 @@
+(For example, the Windows.h header file #define's many common words that break even the simplest C++
+source code; platform.h reverts some of this damage).
+
+The portable implementations of FilesystemImpl, IoSocketImpl and low-level OsSysCallsImpl using
+os_fd_t descriptors and the error constants defined in `platform.h` should be used in favor of raw
+POSIX APIs when authoring tests.
+
+## Dependency behavioral differences
+
+The libraries that Envoy depends on will exhibit different behaviors when built for Windows
+versus Linux versus BSD architectures. In some cases, Envoy has patched the library to either
+correct misbehavior, or to make the library follow expected linux conventions more closely,
+and every attempt is made to push such changes as optional features to upstream. Some notable
+discrepancies include;
+
+The googletest library has a number of Windows-specific quirks. At the time of this writing,
+this library is incompatible with the LLVM clang-cl. It uses the POSIX re (regular expression)
+system library by default. There is an option to use the PCRE library instead, which the
+project has not adopted. On Windows, it uses a massively simplified expression parser which
+is missing support for most common Re expressions. (The solution to this, substituting the
+Google Re library, is proposed but has not been adopted by the googletest project yet).
+
+The opentracing library is missing a Windows build schema (it relies entirely on automake),
+so all tracing-related Envoy extensions are disabled on Windows.
+
+## Bazel behavior differences
+
+There are several circumstances that bazel will end up misbehaving in one environment or another.
+For example, during RBE CI builds, the environment, paths, mounts etc. may have longer paths or
+otherwise hit limits that aren't observed on the contributor's environment.
+
+Similarly, when building on Windows the tools themselves have different path or command line length
+restrictions than on the typical Linux/clang tool chain. These limits include;
+
+* The individual path elements for -Inclusion (this restricts the total path name lengths available to the project).
+* The length of the embedded command line stored in the debug info record of the COFF object file (48kb on Windows).
+
+Note that the rules_foreign_cc macros for bazel helps to ensure compilation is consistent across
+architectures. For the greatest portability, we rely on native bazel BUILD rules, or CMakeLists.txt
+input to CMake which is handled somewhat painlessly by rules_foreign_cc.
+
+Many tests rely on command line scripting or tool invocation. Bazel and Envoy rely heavily on bash scripts, executing
+on Windows via msys2. Inherent discrepancies between the msys2 execution environment and a typical bash shell can
+cause confusing errors. Be sure to use cmd.exe on Windows in any test scripts that intend to create symlinks.
```

Comment:
Should we have a separate section for testing and add there:

1. cmd for symlinks
2. googletest regex
3. googletest nice mocks on base classes

Reply:
Sub-sections, yes as noted above.